internal/security/secretstore/password.go

```diff
@@ -0,0 +1,216 @@
+}
+
+func (gk GokeyGenerator) Generate(realm string) (string, error) {
+	passSpec := gokey.PasswordSpec{
+		Length:         16,
+		Upper:          3,
+		Lower:          3,
+		Digits:         2,
+		Special:        1,
+		AllowedSpecial: "",
+	}
+	t, err := GetAccessToken(gk.tokenPath)
+	if err != nil {
+		return "", err
+	}
+	return gokey.GetPass(t, realm, nil, &passSpec)
+}
+
+type CredCollect struct {
+	Pair UserPasswordPair `json:"data"`
+}
+
+type UserPasswordPair struct {
+	User     string `json:"user,omitempty"`
+	Password string `json:"password,omitempty"`
+}
+
+type Cred struct {
+	client    internal.HttpCaller
+	tokenPath string
+	generator CredentialGenerator
+}
+
+func NewCred(caller internal.HttpCaller, tpath string, generator CredentialGenerator) Cred {
+	return Cred{client: caller, tokenPath: tpath, generator: generator}
+}
+
+func (cr *Cred) AlreadyInStore(path string) (bool, error) {
+	pair, err := cr.getUserPasswordPair(path)
+	if err != nil {
+		if err == errNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	if len(pair.User) > 0 && len(pair.Password) > 0 {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (cr *Cred) getUserPasswordPair(path string) (*UserPasswordPair, error) {
+	t, err := GetAccessToken(cr.tokenPath)
+	if err != nil {
+		return nil, err
+	}
+	pair, err := cr.retrieve(t, path)
+	if err != nil {
+		return nil, err
+	}
+	return pair, nil
+}
+
+func (cr *Cred) retrieve(t string, path string) (*UserPasswordPair, error) {
+	credUrl, err := cr.credPathURL(path)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest(http.MethodGet, credUrl, nil)
+	if err != nil {
+		e := fmt.Errorf("error creating http request: %v", err.Error())
+		LoggingClient.Error(e.Error())
+		return nil, e
+	}
+
+	req.Header.Set(VaultToken, t)
+	resp, err := cr.client.Do(req)
+	if err != nil {
+		e := fmt.Errorf("failed to retrieve the credential pair on path %s with error %s", path, err.Error())
+		LoggingClient.Error(e.Error())
+		return nil, e
+	}
+	defer resp.Body.Close()
+
+	cred := CredCollect{}
+
+	if resp.StatusCode == http.StatusNotFound {
+		LoggingClient.Info(fmt.Sprintf("credential pair NOT found in secret store @/%s, status: %s", path, resp.Status))
+		return nil, errNotFound
+	} else if resp.StatusCode != http.StatusOK {
+		e := fmt.Errorf("failed to retrieve the credential pair on path %s with error code %d", path, resp.StatusCode)
+		LoggingClient.Error(e.Error())
+		return nil, e
+	}
+
+	if err = json.NewDecoder(resp.Body).Decode(&cred); err != nil {
+		e := fmt.Errorf("error decoding json response when retrieving credential pair: %s", err.Error())
+		LoggingClient.Error(e.Error())
+		return nil, e
+	}
+
+	return &cred.Pair, nil
+}
+
+func (cr *Cred) credPathURL(path string) (string, error) {
+	baseURL, err := url.Parse(Configuration.SecretService.GetSecretSvcBaseURL())
+	if err != nil {
+		e := fmt.Errorf("error parsing secret-service url.  check server and port properties")
```

Comment:
it would be nice if the error provided here was specific to the actual issue. in general I dislike using tools that tell me something's wrong, especially when the tool has information about what specifically is wrong. i.e. something like:
```go
e := fmt.Sprintf("error parsing secret-service url: %s",err.Error())
```

Reply:
sorry if I was unclear, but I didn't mean for @tsconn23 to address this, rather @jdharms address this in a followup